Don't allow Flush/PhantomLocal to be the head variable in a block in ThreadedCPS

saambarati opened the pull request:
#### 8b320548f9227eebbffd777122c377f56f9f7daa
<pre>
Don&apos;t allow Flush/PhantomLocal to be the head variable in a block in ThreadedCPS
<a href="https://bugs.webkit.org/show_bug.cgi?id=242096">https://bugs.webkit.org/show_bug.cgi?id=242096</a>

Reviewed by Yusuke Suzuki.

Before this patch, we would allow Flush/PhantomLocal to be the node inside the variableAtHead Operands
in a basic block. However, this causes some issues inside of our CFG simplifcations phase.
CFG simplification will look at the variables at the head of the basic blocks that it&apos;s going to
remove to indicate how we should preserve liveness in its predecessor. We would then preserve
liveness using a Flush if the variableAtHead was Flushed, otherwise, we would use a PhantomLocal.

However, the variable at the head might be a PhantomLocal, and it might be a PhantomLocal
over a variable that&apos;s Flushed. However, in our IsFlushed analysis, we never mark
PhantomLocals as IsFlushed, we only mark value producing nodes, Flush, or Phis.
We do this by traversing the Phi data flow graph to propagate IsFlushed. And a Phi
can never transitively point to a PhantomLocal, so it will never indicate if a
variable is flushed.

To fix this, we just make the variableAtHead contain a Phi instead of
a Flush/PhantomLocal in the situation where it used to be a Flush/PhantomLocal.
This Phi is what the Flush/PhantomLocal used to point to. And the compiler is
already prepared for variableAtHead to point to a Phi since that&apos;s
what happens for GetLocal. GetLocal will point to a Phi that&apos;s inside
variableAtHead. And now what&apos;s pointed to by variableAtHead will
properly indicate if the variable is flushed since it&apos;ll be a Phi or a
SetArgument node.

* Source/JavaScriptCore/dfg/DFGCPSRethreadingPhase.cpp:
(JSC::DFG::CPSRethreadingPhase::canonicalizeFlushOrPhantomLocalFor):
(JSC::DFG::CPSRethreadingPhase::canonicalizeLocalsInBlock):
* Source/JavaScriptCore/dfg/DFGGraph.cpp:
(JSC::DFG::Graph::dumpBlockHeader):

Canonical link: <a href="https://commits.webkit.org/252192@main">https://commits.webkit.org/252192@main</a>
</pre>

Constellation reviewed (approved):
OK, I think I understand what is happening (miscomputation because of non-tracing of IsFlush state from PhantomLocal), and the solution (using already-inserted Phi for head variable (PhantomLocal(Phi))'s Phi, which is correctly traced already.)
r=me

saambarati committed:
Don't allow Flush/PhantomLocal to be the head variable in a block in ThreadedCPS
https://bugs.webkit.org/show_bug.cgi?id=242096

Reviewed by Yusuke Suzuki.

Before this patch, we would allow Flush/PhantomLocal to be the node inside the variableAtHead Operands
in a basic block. However, this causes some issues inside of our CFG simplifcations phase.
CFG simplification will look at the variables at the head of the basic blocks that it's going to
remove to indicate how we should preserve liveness in its predecessor. We would then preserve
liveness using a Flush if the variableAtHead was Flushed, otherwise, we would use a PhantomLocal.

However, the variable at the head might be a PhantomLocal, and it might be a PhantomLocal
over a variable that's Flushed. However, in our IsFlushed analysis, we never mark
PhantomLocals as IsFlushed, we only mark value producing nodes, Flush, or Phis.
We do this by traversing the Phi data flow graph to propagate IsFlushed. And a Phi
can never transitively point to a PhantomLocal, so it will never indicate if a
variable is flushed.

To fix this, we just make the variableAtHead contain a Phi instead of
a Flush/PhantomLocal in the situation where it used to be a Flush/PhantomLocal.
This Phi is what the Flush/PhantomLocal used to point to. And the compiler is
already prepared for variableAtHead to point to a Phi since that's
what happens for GetLocal. GetLocal will point to a Phi that's inside
variableAtHead. And now what's pointed to by variableAtHead will
properly indicate if the variable is flushed since it'll be a Phi or a
SetArgument node.

* Source/JavaScriptCore/dfg/DFGCPSRethreadingPhase.cpp:
(JSC::DFG::CPSRethreadingPhase::canonicalizeFlushOrPhantomLocalFor):
(JSC::DFG::CPSRethreadingPhase::canonicalizeLocalsInBlock):
* Source/JavaScriptCore/dfg/DFGGraph.cpp:
(JSC::DFG::Graph::dumpBlockHeader):

Canonical link: https://commits.webkit.org/252192@main
